doc: add more literature to api doc

tripledes opened the pull request:
After our chat over email (@hardys) , I'm sending this PR that adds:

* diagram in SVG for future modifications
* diagram in PNG for linking
* Markdown document with the explanation and showing the diagram

I could also provide the original Google Drawing if that's preferred over the SVG.

Let me know if anything would need amending or changed completely :nerd_face: 

**UPDATE**:
* Tried to merge both api.md and my previous addition
* Add missing or erroneous fields
* Add a complete example of the BareMetalHost resource

nordixinfra commented:
Can one of the admins verify this patch?

zaneb commented:
It would be good if you could add something to the Makefile for automatically converting the SVG to the PNG.

TBH I'm not sure I see the value of the diagram, since it is not generated from the code it will quickly become out of date. The docs look good, although a lot of the information seems duplicated by https://github.com/metal3-io/baremetal-operator/blob/master/docs/api.md

tripledes commented:
@zaneb my original intention making the diagram and the doc wasn't really submitting a PR but as mentioned, @hardys suggested to send it over. I completely agree this should be kept up to date as it will stale quickly and if it's already (almost) covered makes no sense to duplicate content. Please feel free to drop the PR :+1:

russellb commented:
The current api.md doc has a lot of field reference info, but not so much of the general explanation of the objects.  Can you look at how you might integrate your new content into that file?

As for the diagram, it's nice, but I'd agree with the concern about keeping it up to date.  Maybe split that out into a new PR and we can think about ways to auto generate something similar?  or just drop it for now ...

tripledes commented:
@russellb makes sense, I think I'll drop the diagram for now and give it a thought on how to auto-generate it. I'll see how to integrate the doc on `api.md` and update the PR, thanks!

tripledes commented:
@zaneb yeah, I'm still polishing it a bit, I'll remove the WIP (probably today by the EOD) and ping you for review. Thanks!

tripledes commented:
Hi! I tried to merge both but respecting `api.md` to the best of my knowledge on the language and the code :)

Let me know if you find anything missing (I'm aware I skipped few things, hopefully not that important) or anything wrong....I'll be more than happy to fix/add/remove anything.

zaneb commented:
One issue with this is that you've changed the names from the ones that appear in the API (which start with lower case) to field names or sometimes types (e.g. BareMetalHostStatus) that appear only in the Golang code. Since this is the API document I think we should continue to use the names as they appear in JSON.
The extra details and hierarchy are fantastic though, this is a lot more readable now that the API has grown quite a bit.

tripledes commented:
@zaneb makes sense, I understood the API as the package API, but I get your point, let me fix it and re-push it.

tripledes commented:
@zaneb I reverted all the names and changed the references to `BareMetalHostSpec` and `BareMetalHostStatus`, let me know if that'd be enough. I used the json annotations on the types file to make the modifications.

dhellmann commented on docs/api.md:
```diff
@@ -2,202 +2,266 @@
+        for example `redfish://myhost.example/redfish/v1/Systems/MySystemExample`
+
+  * *credentialsName* -- A reference to a *secret* containing the connection
+    details, at least username and password, for the BMC.
```
```suggestion
    username and password for the BMC.
```

We only want the username and password, so saying "at least" is confusing here.

dhellmann commented on docs/api.md:
```diff
@@ -2,202 +2,266 @@
+* *consumerRef* -- A reference to another resource that is using the
+  host, it could be empty if the host is not being currently used.
+  For example, a *Machine* resource when the host is being used by the
+  [*machine-api*](https://github.com/openshift/machine-api-operator).
```
```suggestion
  [*machine-api*](https://github.com/kubernetes-sigs/cluster-api).
```

We should link to the upstream definition of the machine-api.

dhellmann commented on docs/api.md:
```diff
@@ -2,202 +2,266 @@
+
+* *externallyProvisioned* -- A boolean indicating whether the host
+  provisioning and deprovisioning are managed externally. When set,
+  if no image is provided, the host's power status and hardware
```
```suggestion
  the host's power status and hardware
```

It looks like I missed a doc update when changing the way this flag works. If it is set, the image isn't checked at all.

dhellmann commented on docs/api.md:
```diff
@@ -2,202 +2,266 @@
+  * *checksum* -- The actual checksum or a URL to a file containing
+    the checksum for the image at *image.url*.
+
+* *userData* -- A reference to the Secret containing the user provided data
```
```suggestion
* *userData* -- A reference to the Secret containing the cloudinit user data
```

It's not "user provided data" it's the user data value for cloudinit.

tripledes replied:
@dhellmann I used a kinda generic term cause I assumed it could be other technologies, like for instance ignition, is not that the case?

dhellmann replied:
For RHCOS the data is consumed by Ignition, but it's passed in cloudinit format.

dhellmann commented on docs/api.md:
```diff
@@ -2,202 +2,266 @@
+#### Status fields
+
+* *goodCredentials* -- A reference to the secret and its namespace
+  holding the last set of credentials the system was able to validate
```
```suggestion
  holding the last set of BMC credentials the system was able to validate
```

dhellmann commented on docs/api.md:
```diff
@@ -2,202 +2,266 @@
+  as working.
+
+* *triedCredentials* -- A reference to the secret and its namespace
+  holding the last set of credentials that were sent to
```
```suggestion
  holding the last set of BMC credentials that were sent to
```

dhellmann commented on docs/api.md:
```diff
@@ -2,202 +2,266 @@
+  * *OK* -- Indicates all the details for the host are known and working,
+    meaning the host is correctly configured and manageable.
+  * *discovered* -- Implies some of the host's details are either
+    not working correctly missing. For exmample, the BMC address is known
```
```suggestion
    not working correctly or missing. For example, the BMC address is known
```

dhellmann commented on docs/api.md:
```diff
@@ -2,202 +2,266 @@
+  hardwareProfile: default
+  image:
+    checksum: http://172.16.1.100/images/myOS.qcow2.md5sum
+    url: http://172.16.1.100/images/myOSv1/myOS.qcow2
```
The paths in these values don't match. Should they?

tripledes replied:
I guess they don't have to, but it was my intention that they'd match, fixing it, thanks!

tripledes commented:
@dhellmann I made all the modifications. Let me know if there is anything else!

zaneb commented:
/lgtm
